src/main/java/org/opensearch/flowframework/workflow/WorkflowProcessSorter.java

```diff
@@ -153,31 +153,45 @@ public void validate(List<ProcessNode> processNodes) throws Exception {
      * @throws Exception on validation failure
      */
     public void validatePluginsInstalled(List<ProcessNode> processNodes, WorkflowValidator validator) throws Exception {
-
-        // Retrieve node information to ascertain installed plugins
-        NodesInfoRequest nodesInfoRequest = new NodesInfoRequest();
-        nodesInfoRequest.clear().addMetric(NodesInfoRequest.Metric.PLUGINS.metricName());
-        CompletableFuture<List<String>> installedPluginsFuture = new CompletableFuture<>();
-        client.admin().cluster().nodesInfo(nodesInfoRequest, ActionListener.wrap(response -> {
-            List<String> installedPlugins = new ArrayList<>();
-
-            // Retrieve installed plugin names from the local node
-            String localNodeId = clusterService.state().getNodes().getLocalNodeId();
-            NodeInfo info = response.getNodesMap().get(localNodeId);
-            PluginsAndModules plugins = info.getInfo(PluginsAndModules.class);
-            for (PluginInfo pluginInfo : plugins.getPluginInfos()) {
-                installedPlugins.add(pluginInfo.getName());
-            }
-
-            installedPluginsFuture.complete(installedPlugins);
-
-        }, exception -> {
-            logger.error("Failed to retrieve installed plugins");
-            installedPluginsFuture.completeExceptionally(exception);
+        final CompletableFuture<List<String>> installedPluginsFuture = new CompletableFuture<>();
+
+        ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
+        clusterStateRequest.clear().nodes(true).local(true);
+        client.admin().cluster().state(clusterStateRequest, ActionListener.wrap(stateResponse -> {
```

Comment:
Can we use `  String localNodeId = clusterService.state().getNodes().getLocalNodeId()` instead of sending a cluster state request?

Reply:
Of note, `clusterService.localNode()` internally calls that whole chain and is shorter, but...

What you wrote is what the existing code was, and was possibly causing a deadlock in a multinode cluster.  I'm not sure exactly the cause yet, but this seems the most suspicious.  The `state()` call coded as an `AtomicReference` meaning it has to contend with locking, etc. of the state by the update applier function to be sure the most recent state is returned, all of which depends on a ton of extra information on the cluster that we don't need.

Since all we need is the local node id, we can restrict the cluster state request to `local()` so it fetches from the local node and does not require to fetch from the cluster manager (or have the current state of all the nodes updated). We also restrict it to `nodes()` to only get the nodes info, for a shorter response.

Finally, the way I wrote it is how it's implemented in `_cat/plugins`, so given the unknowns with the `state()` method I feel it's safer.

Reply:
Ah good points, this makes sense to me